server/app/org/scalaexercises/exercises/controllers/ApplicationController.scala

```diff
@@ -12,131 +14,126 @@
-import cats.free._
-import cats.instances.future._
-import play.api.cache.CacheApi
-import org.scalaexercises.types.exercises.{Contribution, Contributor}
-
-import scala.collection.JavaConverters._
-import org.scalaexercises.exercises.utils.ConfigUtils
-import org.scalaexercises.algebra.app._
-import org.scalaexercises.algebra.user.UserOps
-import org.scalaexercises.algebra.progress.{UserExercisesProgress, UserProgressOps}
+import cats.effect.IO
+import cats.implicits._
 import org.scalaexercises.algebra.exercises.ExerciseOps
-import org.scalaexercises.types.github.Repository
 import org.scalaexercises.algebra.github.GithubOps
+import org.scalaexercises.algebra.progress.{UserExercisesProgress, UserProgressOps}
+import org.scalaexercises.algebra.user.UserOps
+import org.scalaexercises.exercises.Secure
 import org.scalaexercises.exercises.services.ExercisesService
-import org.scalaexercises.exercises.services.interpreters.ProdInterpreters
-import doobie.imports._
-import play.api.{Application, Logger, Play}
-import play.api.libs.concurrent.Execution.Implicits.defaultContext
+import org.scalaexercises.exercises.utils.ConfigUtils
+import org.scalaexercises.types.exercises.{Contribution, Contributor}
+import org.scalaexercises.types.github.Repository
+import play.api.cache.AsyncCacheApi
 import play.api.mvc._
 import play.api.routing.JavaScriptReverseRouter
-
-import scala.concurrent.Future
-import scala.concurrent.duration._
-import scalaz.concurrent.Task
-import freestyle._
-import freestyle.implicits._
-
-class ApplicationController(cache: CacheApi)(
-    implicit exerciseOps: ExerciseOps[ExercisesApp.Op],
-    userOps: UserOps[ExercisesApp.Op],
-    userProgressOps: UserProgressOps[ExercisesApp.Op],
-    userExercisesProgress: UserExercisesProgress[ExercisesApp.Op],
-    githubOps: GithubOps[ExercisesApp.Op],
-    T: Transactor[Task]
-) extends Controller
-    with AuthenticationModule
-    with ProdInterpreters {
-  implicit def application: Application = Play.current
-
-  lazy val topLibraries: List[String] = application.configuration.getStringList(
-    "exercises.top_libraries") map (_.asScala.toList) getOrElse Nil
+import play.api.{Configuration, Logger, Mode}
+
+class ApplicationController(
+    config: Configuration,
+    cl: ClassLoader,
+    components: ControllerComponents)(cache: AsyncCacheApi)(
+    implicit exerciseOps: ExerciseOps[IO],
+    userOps: UserOps[IO],
+    userProgressOps: UserProgressOps[IO],
+    userExercisesProgress: UserExercisesProgress[IO],
+    githubOps: GithubOps[IO],
+    mode: Mode)
+    extends BaseController
+    with AuthenticationModule {
+  lazy val topLibraries: List[String] = config.getOptional[Seq[String]]("exercises.top_libraries") map (_.toList) getOrElse Nil
 
   val MainRepoCacheKey = "scala-exercises.repo"
 
+  val configUtils = ConfigUtils(config)
+
+  private val service = new ExercisesService(cl)
+
+  private implicit val cs = IO.contextShift(defaultExecutionContext)
```

Comment:
This comment applies to the entire application.

Could we pass this kind of instances (ExecutionContexts, ContextShifts, Timers, etc.) implicitly to every controller and service? So we are building just one instance of every time once, and only at the beginning (when the application starts).

Reply:
It has sense when we're talking about configuration, application mode... to be a val. Should I declare the rest (ExecutionContexts, Timers...) as a def instead val?